src/test/neorados/common_tests.h

```diff
@@ -12,30 +12,286 @@
+  }
+
+  /// \brief TestBody coroutine
+  ///
+  /// Run after setup.
+  virtual boost::asio::awaitable<void> CoTestBody() = 0;
+
+  /// \brief Run our coroutines
+  ///
+  /// This is marked final, since the actual test body belongs in
+  /// CoTestBody.
+  ///
+  /// Run CoSetUp and, if CoSetUp succeeded, CoTestBody and
+  /// CoTearDown.
+  ///
+  /// Error reporting of failures in CoSetUp and CoTearDown leaves
+  /// something to be desired as GTest thinks everything is the test
+  /// proper.
+  void TestBody() override final {
+    boost::asio::co_spawn(
+      asio_context,
+      [](CoroTest* t) -> boost::asio::awaitable<void> {
+	co_await t->CoSetUp();
+	try {
+	  co_await t->CoTestBody();
+	} catch (...) {
+	  t->eptr = std::current_exception();
+	}
+	co_await t->CoTearDown();
+	if (t->eptr) {
+	  std::rethrow_exception(t->eptr);
+	}
+	co_return;
+      }(this),
+      [](std::exception_ptr e) {
+	if (e) std::rethrow_exception(e);
+      });
+    asio_context.run();
+  }
+};
+
+/// \brief C++20 coroutine test harness for NeoRados
+///
+/// CoTestBody has access to `rados`, a `neorados::RADOS` handle, and
+/// `pool`, a `neorados::IOContext` representing a pool that will be
+/// destroyed when the test exits.
+class NeoRadosTest : public CoroTest {
+private:
+  /// Workaround so we don't have to use an optional or allocate or
+  /// something.
+  alignas(neorados::RADOS) std::byte rspace[sizeof(neorados::RADOS)];
+protected:
+  neorados::RADOS& rados =
+    *reinterpret_cast<neorados::RADOS*>(rspace);
+  neorados::IOContext pool; ///< The pool usable by this test.
+public:
+
+  /// \brief Create RADOS handle and pool for the test
+  boost::asio::awaitable<void> CoSetUp() override {
+    auto r = co_await neorados::RADOS::Builder{}
+      .build(asio_context, boost::asio::use_awaitable);
+    new (&rados) neorados::RADOS(std::move(r));
+    pool.pool(co_await create_pool(rados, get_temp_pool_name(
+				     testing::UnitTest::GetInstance()
+				     ->current_test_info()->name()),
+				   boost::asio::use_awaitable));
+    co_return;
+  }
+
+  ~NeoRadosTest() override {
+    rados.~RADOS();
+  }
+
+  /// \brief Delete pool used for testing
+  boost::asio::awaitable<void> CoTearDown() override {
+    co_await rados.delete_pool(pool.pool(),
+				boost::asio::use_awaitable);
+    co_return;
+  }
+};
+
+/// \brief Helper macro for defining coroutine tests with a fixture
+///
+/// Defines a test using a coroutine fixture for
+/// SetUp/TearDown. Fixtures must be descendants of `CoroTest`.
+///
+/// \note Uses more of GTest's internals that I would like.
+///
+/// \warning Use `EXPECT_*` only, not `ASSERT_*`. `ASSERT_` macros
+/// return from the calling function and will not work in a
+/// coroutine.
+///
+/// \param test_suite_name Name of the test suite
+/// \param test_name Name of the test
+/// \param fixture Fixture class to use (descendent of CoroTest)
+#define CORO_TEST_F(test_suite_name, test_name, fixture)                       \
```

Comment:
this is neat, and could be useful outside of neorados too

i'm a little concerned about the reliance on `testing::internal`, which could make it harder to support different versions of gtest. afaik we're still using submoduled gtest, so maybe it's not a big issue?

Reply:
I couldn't really find a way to avoid it. I thought of re-implement the internal call myself, but I wanted to avoid that since it depended on the guts of the thing but less explicitly.

I think since we're using a submodule it should be okay.